File: 1_Algorithmic_Toolbox/week4_divide_and_conquer/1_binary_search/binary_search.cpp
#include <iostream>
#include <cassert>
#include <vector>

using std::vector;

using namespace std;

int binary_search_helper(const vector<int> &a, int left, int right, int key) {

  if (left > right) {
    return -1;
  }

  int middle_index = (left + right) / 2;

  if (a[middle_index] == key) {
    return middle_index;
  } else if (a[middle_index] > key) {
    return binary_search_helper(a, left, middle_index-1, key);
  } else if (a[middle_index] < key) {
    return binary_search_helper(a, middle_index+1, right, key);
  }
  return -1;
}

int binary_search(const vector<int> &a, int x) {
  //write your code here
  return binary_search_helper(a, 0, (int)a.size()-1, x);
}

int linear_search(const vector<int> &a, int x) {
  for (size_t i = 0; i < a.size(); ++i) {
    if (a[i] == x) return i;
  }
  return -1;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> a(n);
  for (size_t i = 0; i < a.size(); i++) {
    std::cin >> a[i];
  }
  int m;
  std::cin >> m;
  vector<int> b(m);
  for (int i = 0; i < m; ++i) {
    std::cin >> b[i];
  }
  for (int i = 0; i < m; ++i) {
    //replace with the call to binary_search when implemented
    // std::cout << linear_search(a, b[i]) << ' ';
    std::cout << binary_search(a, b[i]) << ' ';
  }
}

File: 1_Algorithmic_Toolbox/week4_divide_and_conquer/1_binary_search/binary_search.py
# Uses python3
import sys


def binary_search_helper(a, left, right, x):
    if left >= right:
        return -1

    # write your code here
    middle_index = (left + right) // 2

    if (a[middle_index] == x):
        return middle_index
    elif (a[middle_index] > x):
        return binary_search_helper(a, left, middle_index, x)
    elif (a[middle_index] < x):
        return binary_search_helper(a, middle_index+1, right, x)
    return -1


def binary_search(a, x):
    return binary_search_helper(a, 0, len(a)-1, x)


def linear_search(a, x):
    for i in range(len(a)):
        if a[i] == x:
            return i
    return -1


def main():
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n = data[0]
    a = data[1: n + 1]
    for x in data[n + 2:]:
        # replace with the call to binary_search when implemented
        # print(linear_search(a, x), end = ' ')
        print(binary_search(a, x), end=' ')


if __name__ == '__main__':
    main()
